File: bench/compare/go/sorted_lookup.go
package main

type entry struct{ key, value int64 }

func cmpI64(a, b int64) int64 {
	if a < b {
		return -1
	}
	if a > b {
		return 1
	}
	return 0
}

func buildMap(n int64) []entry {
	m := make([]entry, 0, n)
	for i := int64(0); i < n; i++ {
		m = append(m, entry{i * 2, i})
	}
	return m
}

func find(m []entry, key int64, cmp func(int64, int64) int64) (int64, bool) {
	lo, hi := int64(0), int64(len(m))
	for lo < hi {
		mid := lo + (hi-lo)/2
		c := cmp(key, m[mid].key)
		if c == 0 {
			return mid, true
		}
		if c < 0 {
			hi = mid
		} else {
			lo = mid + 1
		}
	}
	return lo, false
}

func lookupSweep(m []entry, span int64, cmp func(int64, int64) int64) int64 {
	var sum int64
	for k := int64(0); k < span; k++ {
		if idx, ok := find(m, k, cmp); ok {
			sum += m[idx].value
		} else {
			sum++
		}
	}
	return sum
}

func main() {
	n := int64(20000)
	reps := 10
	m := buildMap(n)
	var total int64
	for r := 0; r < reps; r++ {
		total += lookupSweep(m, n*2, cmpI64)
	}
	if total != 2000100000 {
		panic(total)
	}
}

File: bench/compare/rust/sorted_lookup.rs
struct Entry {
    key: i64,
    value: i64,
}

fn cmp_i64(a: i64, b: i64) -> i64 {
    if a < b {
        -1
    } else if a > b {
        1
    } else {
        0
    }
}

fn build_map(n: i64) -> Vec<Entry> {
    let mut m = Vec::with_capacity(n as usize);
    for i in 0..n {
        m.push(Entry { key: i * 2, value: i });
    }
    m
}

fn find(m: &[Entry], key: i64, cmp: impl Fn(i64, i64) -> i64) -> (i64, bool) {
    let mut lo: i64 = 0;
    let mut hi: i64 = m.len() as i64;
    while lo < hi {
        let mid = lo + (hi - lo) / 2;
        let c = cmp(key, m[mid as usize].key);
        if c == 0 {
            return (mid, true);
        }
        if c < 0 {
            hi = mid;
        } else {
            lo = mid + 1;
        }
    }
    (lo, false)
}

fn lookup_sweep(m: &[Entry], span: i64) -> i64 {
    let mut sum: i64 = 0;
    for k in 0..span {
        let (idx, ok) = find(m, k, cmp_i64);
        if ok {
            sum += m[idx as usize].value;
        } else {
            sum += 1;
        }
    }
    sum
}

fn main() {
    let n: i64 = 20000;
    let reps = 10;
    let m = build_map(n);
    let mut total: i64 = 0;
    for _ in 0..reps {
        total += lookup_sweep(&m, n * 2);
    }
    if total != 2000100000 {
        panic!("{}", total);
    }
}
